### ISSUE TITLE
loading 32 bit TIFF requires imread flags to be set

### ISSUE THREAD

**Issue by mixmixmix** [opencv/opencv#7025]
According to the documentation flag `IMREAD_ANYDEPTH ` 

> If set, return 16-bit/32-bit image when the input has the corresponding depth, otherwise convert it to 8-bit

When reading 32bit TIFF image without any flags the image should load and get converted to 8bit. In practice however image is empty.

**Comment by Asalle** [comment]
@StevenPuttemans 
Okay, I need a little help with it.
I looked at imread_ (the implementation) and have not found any check for the IMREAD_ANYDEPTH flag whatsoever. Also, as far as I understood, the ImageDecoder class does almost all the job and this flag has to be transferred to it, just like other flags, e. g. IMREAD_REDUCED_GRAYSCALE_2, but there is no setter function for this flag. Then, how using this flag is even possible?

**Comment by mshabunin** [comment]
@Asalle , please take a look at the `TiffDecoder::readData( Mat& img )` function. It should fill incoming image with correct data, but for some reason in case when actual TIFF has 32-bit depth and target image 8-bit depth, the decoding process fails.

**Comment by Asalle** [comment]
I'm assuming it affected master

**Comment by mshabunin** [comment]
That's ok, actual generated code and execution order does not necessary match original source code line-by-line. However make sure you are using **debug build** to minimize compiler optimizations: `-DCMAKE_BUILD_TYPE=Debug` cmake parameter.

**Comment by Asalle** [comment]
Sure thing I do

**Comment by mshabunin** [comment]
@Asalle , actually debugging OpenCV is a great idea. I'd recommend you to use gcc + gdb + QtCreator for visual debugging.

**Comment by StevenPuttemans** [comment]
I agree, debugging is indeed a good way to dive into code and add breaks where needed to see if the correct data is being passed on. I actually use it all the time to discover the nature of specific issues.

**Comment by Asalle** [comment]
Well, I don't seem to be able to re-create this bug. 
On master: I created a 32-bit tiff (as far as I understand) with the following code
```
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
using namespace cv;
int main()
{
    cv::Mat img(320, 240, CV_32FC3, cv::Scalar(0, 0, 50));
    imwrite("test32bit_red.tiff", img);
}
```
This is just a red rectangle.
Then, I try to read it show it and save under another name, but the image is not empty, it's just in shades of grey:
```
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <cv.h>
using namespace cv;
int main(int argc, char** argv)
{
    Mat img;
    img = imread(argv[1], 0);
    imwrite("test.tiff", img);
    imshow(argv[0], img);
    waitKey(0);
    return 0;
}
```
Input image:
[test32bit_red.tiff.zip](https://github.com/opencv/opencv/files/440796/test32bit_red.tiff.zip)
Output image:
[test.tiff.zip](https://github.com/opencv/opencv/files/440797/test.tiff.zip)

**Comment by StevenPuttemans** [comment]
@Asalle the way you are reading in the image, forces it to be read as a 8 bit grayscale image, the `0` at the `imread` function is the integer value of `IMREAD_GRAYSCALE `. To solve this issue, and be able to read 32 bit images you need to specify this value as `IMREAD_ANYDEPTH `, which is explained [here](http://docs.opencv.org/master/d4/da8/group__imgcodecs.html#ga61d9b0126a3e57d9277ac48327799c80).

Your code transforms then to

```
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <cv.h>
using namespace cv;
int main(int argc, char** argv)
{
    Mat img;
    img = imread(argv[1], IMREAD_ANYDEPTH);
    imwrite("test.tiff", img);
    imshow(argv[0], img);
    waitKey(0);
    return 0;
}
```

Can you try this?

**Comment by saskatchewancatch** [comment]
You can use `tiffinfo` which is part of ["TIFF Tools"](http://www.libtiff.org/tools.html) to get info on the Tiff file. Running `tiffinfo` on test32bit_red.tiff, I get this information:
```
  Image Width: 240 Image Length: 320
  Bits/Sample: 16
  Sample Format: signed integer
  Compression Scheme: SGILog
  Photometric Interpretation: CIE Log2(L) (u',v')
  Samples/Pixel: 3
  Rows/Strip: 1
  Planar Configuration: single image plane
```
In Tiff parlance, a "Sample" is what we'd call a "Channel." That image has an L, a, and b components, 16 bits dedicated to each? Is this the sort of Tiff you expected from `imwrite`?

**Comment by saskatchewancatch** [comment]
I happened to have been toying with Tiffs, and would like to share some findings:

With or without `IMREAD_ANYDEPTH`, `imread` appears to be producing an empty `Mat` when reading in 3 channel, 32 bits per channel Tiffs.

I generated such a Tiff (100x100 w/ random/zero data) in C using libtiff. The source and result image with random/zero image pixel is attached in the zip file.

Run `tiffinfo` (from Tiff-tools) on the image to confirm that it does have 3 channels and 32 bits per channel. Run `tiffdump` on it too to see that it has 100 rows of data.

Load it with `imread` w/ and w/o `IMREAD_ANYDEPTH`, and notice that the resulting `Mat` has no data (debugger shows 0 rows and 0 cols).
[32bps-tiff.zip](https://github.com/opencv/opencv/files/924014/32bps-tiff.zip)
